In [27]:
%reload_ext autoreload
%autoreload 2

In [28]:
from sqlalchemy import create_engine
import numpy as np
import pandas as pd
import matplotlib.pylab as plt
import seaborn as sns
plt.style.use('ggplot')
pd.set_option('display.max_columns', 200) ## view all the columns

In [29]:
from dotenv import load_dotenv
import os

# Load environment variables from .env file
load_dotenv()

True

In [30]:
# connecting to db
db_user = os.getenv("POSTGRES_USER")
db_password = os.getenv("POSTGRES_PASSWORD")
db_port = '5435'  
db_name = os.getenv("POSTGRES_DB")

# Create a connection to the PostgreSQL database
engine = create_engine(f'postgresql://{db_user}:{db_password}@localhost:{db_port}/{db_name}')

# Query to retrieve all rows from the xdr_data
query = "SELECT * FROM xdr_data"

# Execute the query and load the result into a Pandas DataFrame
df = pd.read_sql(query, engine)

# Display the DataFrame
df

Bearer Id            Start  Start ms              End  End ms  \
0       1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1       1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2       1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3       1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4       1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   
...              ...              ...       ...              ...     ...   
149996  7.277826e+18   4/29/2019 7:28     451.0   4/30/2019 6:02   214.0   
149997  7.349883e+18   4/29/2019 7:28     483.0  4/30/2019 10:41   187.0   
149998  1.311448e+19   4/29/2019 7:28     283.0  4/30/2019 10:46   810.0   
149999  1.311448e+19   4/29/2019 7:28     696.0  4/30/2019 10:40   327.0   
150000           NaN             None       NaN             None     NaN   

        Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0       1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1       1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2       1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3       1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4       1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   
...           ...           ...            ...           ...   
149996    81230.0  2.082022e+14   3.365069e+10  3.548311e+13   
149997    97970.0  2.082019e+14   3.366345e+10  3.566051e+13   
149998    98249.0  2.082017e+14   3.362189e+10  3.572121e+13   
149999    97910.0  2.082021e+14   3.361962e+10  8.618620e+13   
150000        NaN           NaN            NaN           NaN   

           Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0       9.16456699548519E+015             42.0              5.0   
1                     L77566A             65.0              5.0   
2                     D42335A              NaN              NaN   
3                     T21824A              NaN              NaN   
4                     D88865A              NaN              NaN   
...                       ...              ...              ...   
149996                D20434A             32.0              0.0   
149997                D10223C             27.0              2.0   
149998                T51102A             43.0              6.0   
149999                L88342B             37.0              5.0   
150000                   None              NaN              NaN   

        Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                          23.0                     44.0   
1                          16.0                     26.0   
2                           6.0                      9.0   
3                          44.0                     44.0   
4                           6.0                      9.0   
...                         ...                      ...   
149996                     52.0                     65.0   
149997                     23.0                     54.0   
149998                     43.0                     47.0   
149999                     34.0                     37.0   
150000                      NaN                      NaN   

        TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                               NaN                          NaN   
1                               NaN                          NaN   
2                               NaN                          NaN   
3                               NaN                          NaN   
4                               NaN                          NaN   
...                             ...                          ...   
149996                          NaN                          NaN   
149997                          NaN                          NaN   
149998                          NaN                          NaN   
149999                          NaN                          NaN   
150000                          NaN                    

## Data Understanding
- Dataframe shape
- head and tail
- dtypes
- describe

In [31]:
df.shape

(150001, 55)

In [32]:
# view top 10 of table
df.head(10)

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   
5  1.311448e+19  4/12/2019 21:37     439.0   4/25/2019 8:08   553.0   
6  1.311448e+19   4/13/2019 8:41     612.0   4/25/2019 8:16   168.0   
7  1.304243e+19   4/14/2019 2:11     592.0   4/25/2019 2:26   512.0   
8  1.311448e+19  4/14/2019 12:48     121.0  4/25/2019 10:22   960.0   
9  1.304243e+19   4/15/2019 0:32       0.0   4/25/2019 0:40   284.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   
5  1074638.0  2.082014e+14   3.366819e+10  3.529841e+13   
6  1035261.0  2.082014e+14   3.366537e+10  8.676270e+13   
7   951292.0  2.082010e+14   3.376349e+10  8.654640e+13   
8   941634.0  2.082014e+14   3.369874e+10  3.556241e+13   
9   864482.0  2.082003e+14   3.365922e+10  3.557311e+13   

      Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0  9.16456699548519E+015             42.0              5.0   
1                L77566A             65.0              5.0   
2                D42335A              NaN              NaN   
3                T21824A              NaN              NaN   
4                D88865A              NaN              NaN   
5                T89132C              NaN              NaN   
6  9.16456701058919E+015            102.0              5.0   
7   CELL_208_20_520025_1             39.0             18.0   
8                T42084A              NaN              NaN   
9                L20434C             97.0              7.0   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   
5                     70.0                     54.0   
6                     22.0                     38.0   
7                   3698.0                    416.0   
8                     46.0                     49.0   
9                   3845.0                   2535.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   
5                          NaN                          NaN   
6                          NaN                          NaN   
7                      19520.0                          NaN   
8                          NaN                          NaN   
9                    3231397.0                       7230.0   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   
5                100.0                             0.0   
6                100.0                             0.0   
7                 99.0                             0.0   
8                100.0                             0.0   
9          

In [33]:
# view the bottom 10 
df.tail(10)

Bearer Id           Start  Start ms              End  End ms  \
149991  7.349883e+18  4/29/2019 7:28     794.0   4/30/2019 0:36   523.0   
149992  1.311448e+19  4/29/2019 7:28     114.0   4/30/2019 7:04   724.0   
149993  1.311448e+19  4/29/2019 7:28      79.0  4/30/2019 18:22   512.0   
149994  1.311448e+19  4/29/2019 7:28      83.0  4/30/2019 15:01   268.0   
149995  1.304243e+19  4/29/2019 7:28     615.0   4/30/2019 0:01   407.0   
149996  7.277826e+18  4/29/2019 7:28     451.0   4/30/2019 6:02   214.0   
149997  7.349883e+18  4/29/2019 7:28     483.0  4/30/2019 10:41   187.0   
149998  1.311448e+19  4/29/2019 7:28     283.0  4/30/2019 10:46   810.0   
149999  1.311448e+19  4/29/2019 7:28     696.0  4/30/2019 10:40   327.0   
150000           NaN            None       NaN             None     NaN   

        Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
149991    61661.0  2.082017e+14   3.376215e+10  3.586061e+13   
149992    84940.0  2.082014e+14   3.376127e+10  3.532701e+13   
149993   125622.0  2.082014e+14   3.362611e+10  3.573531e+13   
149994   113545.0  2.082003e+14   3.364566e+10  3.515541e+13   
149995    59587.0  2.082014e+14   3.366865e+10  3.533251e+13   
149996    81230.0  2.082022e+14   3.365069e+10  3.548311e+13   
149997    97970.0  2.082019e+14   3.366345e+10  3.566051e+13   
149998    98249.0  2.082017e+14   3.362189e+10  3.572121e+13   
149999    97910.0  2.082021e+14   3.361962e+10  8.618620e+13   
150000        NaN           NaN            NaN           NaN   

       Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
149991            D10033B             27.0              2.0   
149992            D78058B             37.0              4.0   
149993            D73542B             46.0              6.0   
149994            T88383B              NaN              NaN   
149995            T85721A            313.0              9.0   
149996            D20434A             32.0              0.0   
149997            D10223C             27.0              2.0   
149998            T51102A             43.0              6.0   
149999            L88342B             37.0              5.0   
150000               None              NaN              NaN   

        Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
149991                     62.0                     54.0   
149992                     23.0                     40.0   
149993                     43.0                     41.0   
149994                     55.0                     54.0   
149995                  63420.0                   1393.0   
149996                     52.0                     65.0   
149997                     23.0                     54.0   
149998                     43.0                     47.0   
149999                     34.0                     37.0   
150000                      NaN                      NaN   

        TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
149991                          NaN                          NaN   
149992                          NaN                          NaN   
149993                          NaN                          NaN   
149994                          NaN                          NaN   
149995                   16552848.0                     162614.0   
149996                          NaN                          NaN   
149997                          NaN                          NaN   
149998                          NaN                          NaN   
149999                          NaN                          NaN   
150000                          NaN                          NaN   

        DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
149991                100.0                             0.0   
149992                100.0                             0.0   
149993                100.0                             0.0   
149994                100.0                             0.0   
149995                 54.0                            23.0  

In [34]:
# view the all columns data types for this table 
df.dtypes

Bearer Id                                   float64
Start                                        object
Start ms                                    float64
End                                          object
End ms                                      float64
Dur. (ms)                                   float64
IMSI                                        float64
MSISDN/Number                               float64
IMEI                                        float64
Last Location Name                           object
Avg RTT DL (ms)                             float64
Avg RTT UL (ms)                             float64
Avg Bearer TP DL (kbps)                     float64
Avg Bearer TP UL (kbps)                     float64
TCP DL Retrans. Vol (Bytes)                 float64
TCP UL Retrans. Vol (Bytes)                 float64
DL TP < 50 Kbps (%)                         float64
50 Kbps < DL TP < 250 Kbps (%)              float64
250 Kbps < DL TP < 1 Mbps (%)               float64
DL TP > 1 Mb

In [35]:
# over view of the table
df.describe()

Bearer Id       Start ms         End ms     Dur. (ms)          IMSI  \
count  1.490100e+05  150000.000000  150000.000000  1.500000e+05  1.494310e+05   
mean   1.013887e+19     499.188200     498.800880  1.046086e+05  2.082016e+14   
std    2.893173e+18     288.611834     288.097653  8.103762e+04  2.148809e+10   
min    6.917538e+18       0.000000       0.000000  7.142000e+03  2.040471e+14   
25%    7.349883e+18     250.000000     251.000000  5.744050e+04  2.082014e+14   
50%    7.349883e+18     499.000000     500.000000  8.639900e+04  2.082015e+14   
75%    1.304243e+19     749.000000     750.000000  1.324302e+05  2.082018e+14   
max    1.318654e+19     999.000000     999.000000  1.859336e+06  2.140743e+14   

       MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
count   1.489350e+05  1.494290e+05    122172.000000    122189.000000   
mean    4.188282e+10  4.847455e+13       109.795706        17.662883   
std     2.447443e+12  2.241637e+13       619.782739        84.793524   
min     3.360100e+10  4.400152e+11         0.000000         0.000000   
25%     3.365130e+10  3.546071e+13        32.000000         2.000000   
50%     3.366371e+10  3.572201e+13        45.000000         5.000000   
75%     3.368349e+10  8.611970e+13        70.000000        15.000000   
max     8.823971e+14  9.900120e+13     96923.000000      7120.000000   

       Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
count            150000.000000            150000.000000   
mean              13300.045927              1770.428647   
std               23971.878541              4625.355500   
min                   0.000000                 0.000000   
25%                  43.000000                47.000000   
50%                  63.000000                63.000000   
75%               19710.750000              1120.000000   
max              378160.000000             58613.000000   

       TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
count                 6.185500e+04                 5.335200e+04   
mean                  2.080991e+07                 7.596587e+05   
std                   1.825665e+08                 2.645305e+07   
min                   2.000000e+00                 1.000000e+00   
25%                   3.565150e+04                 4.694750e+03   
50%                   5.687300e+05                 2.094950e+04   
75%                   3.768308e+06                 8.402025e+04   
max                   4.294426e+09                 2.908226e+09   

       DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
count        149247.000000                   149247.000000   
mean             92.844754                        3.069355   
std              13.038031                        6.215233   
min               0.000000                        0.000000   
25%              91.000000                        0.000000   
50%             100.000000                        0.000000   
75%             100.000000                        4.000000   
max             100.000000                       93.000000   

       250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
count                  149247.000000       149247.000000        149209.000000   
mean                        1.717341            1.609654            98.530142   
std                         4.159538            4.828890             4.634285   
min                         0.000000            0.000000             0.000000   
25%                         0.000000            0.000000            99.000000   
50%                         0.000000            0.000000           100.000000   
75%                         1.000000            0.000000           100.000000   
max                       100.000000           94.000000           100.000000   

       10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
count                  149209.000000                   149209.000000   
mean                        0.776749                        0.

### For the actual telecom dataset, you‘re expected to conduct a full User Overview analysis & the following sub-tasks are your guidance: 
- Start by identifying the top 10 handsets used by the customers.
- Then, identify the top 3 handset manufacturers
- Next, identify the top 5 handsets per top 3 handset manufacturer
- Make a short interpretation and recommendation to marketing teams


In [11]:
# top 10 headsets used by the customer
top_10_handset_by_customers= df['Handset Type'].value_counts()[:10]
top_10_handset_by_customers

Huawei B528S-23A                19752
Apple iPhone 6S (A1688)          9419
Apple iPhone 6 (A1586)           9023
undefined                        8987
Apple iPhone 7 (A1778)           6326
Apple iPhone Se (A1723)          5187
Apple iPhone 8 (A1905)           4993
Apple iPhone Xr (A2105)          4568
Samsung Galaxy S8 (Sm-G950F)     4520
Apple iPhone X (A1901)           3813
Name: Handset Type, dtype: int64

In [36]:
# identify the top 3 headset manufactures 
top_3_handset_by_manufactures = df['Handset Manufacturer'].value_counts()[:3]
top_3_handset_by_manufactures

Apple      59565
Samsung    40839
Huawei     34423
Name: Handset Manufacturer, dtype: int64

In [37]:
# so totall of users that use those brands will be 59565+40839+34423 = 134827

In [26]:
# identify the top 5 handsets per top 3 handset manufacturer
# manufactor one : Apple 
# Filter data for Apple handsets
apple_top_handsets = df.loc[df['Handset Manufacturer'] == 'Apple']

# Group by Handset Manufacturer and Handset Type, then aggregate the count
apple_top_handsets = apple_top_handsets.groupby(['Handset Manufacturer', 'Handset Type']).agg(count=('Handset Type', 'count'))

# Select the top 5 handsets based on count
apple_top_handsets = apple_top_handsets.nlargest(5, 'count')

# Display the result
print(apple_top_handsets)

                                              count
Handset Manufacturer Handset Type                  
Apple                Apple iPhone 6S (A1688)   9419
                     Apple iPhone 6 (A1586)    9023
                     Apple iPhone 7 (A1778)    6326
                     Apple iPhone Se (A1723)   5187
                     Apple iPhone 8 (A1905)    4993


In [25]:
# identify the top 5 handsets per top 3 handset manufacturer
# manufactor two : Samsung 
# Filter data for Samsung handsets
samsung_top_handsets = df.loc[df['Handset Manufacturer'] == 'Samsung']

# Group by Handset Manufacturer and Handset Type, then aggregate the count
samsung_top_handsets = samsung_top_handsets.groupby(['Handset Manufacturer', 'Handset Type']).agg(count=('Handset Type', 'count'))

# Select the top 5 handsets based on count
samsung_top_handsets = samsung_top_handsets.nlargest(5, 'count')

# Display the result
print(samsung_top_handsets)

                                                   count
Handset Manufacturer Handset Type                       
Samsung              Samsung Galaxy S8 (Sm-G950F)   4520
                     Samsung Galaxy A5 Sm-A520F     3724
                     Samsung Galaxy J5 (Sm-J530)    3696
                     Samsung Galaxy J3 (Sm-J330)    3484
                     Samsung Galaxy S7 (Sm-G930X)   3199


In [24]:
# identify the top 5 handsets per top 3 handset manufacturer
# manufactor three : Huawei 
# Filter data for Huawei handsets
huawei_top_handsets = df.loc[df['Handset Manufacturer'] == 'Huawei']

# Group by Handset Manufacturer and Handset Type, then aggregate the count
huawei_top_handsets = huawei_top_handsets.groupby(['Handset Manufacturer', 'Handset Type']).agg(count=('Handset Type', 'count'))

# Select the top 5 handsets based on count
huawei_top_handsets = huawei_top_handsets.nlargest(5, 'count')

# Display the result
print(huawei_top_handsets)

                                                     count
Handset Manufacturer Handset Type                         
Huawei               Huawei B528S-23A                19752
                     Huawei E5180                     2079
                     Huawei P20 Lite Huawei Nova 3E   2021
                     Huawei P20                       1480
                     Huawei Y6 2018                    997


### Strategic Marketing Insights and Recommendations for Top Smartphone Models

The smartphones manufactured by Apple, Samsung, and Huawei are the most popular among the network's **134,827 users**. The Huawei **B528S-23A** is the most popular device, with **19,752 units** sold. Notably, the Apple iPhone **6S, 6, 7, and SE** take the next four spots in user preferences after Huawei.
For improved market positioning, I would recommend focusing on promoting **Huawei B528S-23A** and its counterparts, as well as strategically targeting **iPhone 6S, 6, 7, and SE**.